okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.java

```diff
@@ -0,0 +1,134 @@
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import javax.annotation.Nullable;
+import okhttp3.Dns;
+import okhttp3.HttpUrl;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.internal.platform.Platform;
+import okio.ByteString;
+
+/**
+ * DNS over HTTPS implementation.
+ *
+ * Implementation of https://tools.ietf.org/html/draft-ietf-doh-dns-over-https-07
+ */
+public class DnsOverHttps implements Dns {
+  public static final MediaType DNS_MESSAGE = MediaType.parse("application/dns-message");
+  public static final MediaType UDPWIREFORMAT = MediaType.parse("application/dns-udpwireformat");
+  private final OkHttpClient client;
+  private final HttpUrl url;
+  private final boolean includeIPv6;
+  private final boolean post;
+  private final MediaType contentType;
+
+  public DnsOverHttps(OkHttpClient client, HttpUrl url,
+      @Nullable Dns bootstrapDns, boolean includeIPv6, String method, MediaType contentType) {
+    this.client = bootstrapDns != null ? client.newBuilder().dns(bootstrapDns).build() : client;
+    this.url = url;
+    this.includeIPv6 = includeIPv6;
+    if (!method.equals("GET") && !method.equals("POST")) {
+      throw new UnsupportedOperationException("Only GET and POST Supported");
+    }
+    this.post = method.equals("POST");
+    this.contentType = contentType;
+  }
+
+  public HttpUrl getUrl() {
+    return url;
+  }
+
+  public boolean isPost() {
+    return post;
+  }
+
+  @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
+    try {
+      //System.out.println("Host: " + hostname);
+
+      ByteString query = DnsRecordCodec.encodeQuery(hostname, includeIPv6);
+
+      Request request = buildRequest(query);
+      Response response = client.newCall(request).execute();
+
+      // TODO reenable (currently noisy with test servers)
+      if (response.cacheResponse() == null && response.protocol() != Protocol.HTTP_2) {
+        Platform.get().log(Platform.WARN, "Incorrect protocol: " + response.protocol(), null);
+      }
+
+      // TODO remove (temporary info only currently)
+      if (client.cache() != null && !post && response.cacheResponse() == null) {
+        Platform.get().log(Platform.INFO, "DNS missed cache: " + hostname, null);
+      }
+
+      try {
+        if (!response.isSuccessful()) {
+          throw new IOException("response: " + response.code() + " " + response.message());
+        }
+
+        ByteString responseBytes = response.body().source().readByteString();
+
+        //System.out.println("Response: " + responseBytes.hex());
+
+        List<InetAddress> results = DnsRecordCodec.decodeAnswers(hostname, responseBytes);
+
+        return results;
+      } finally {
+        response.close();
+      }
+    } catch (UnknownHostException uhe) {
```

Comment:
I don’t think you need this? This only happens when you can’t resolve the DNS server, right?

Reply:
What should this be in this case? 

a) an UnknownHostException for the DNS server (leave this catch block)
b) an UnknownHostException for URL host with a) as the cause
c) other?

Reply:
I think `a` or `b` are both good.

If you're offline you'll experience it as not being able to reach the DNS server. I’d like the DNS server’s name in the message or cause.